### ISSUE TITLE
City UI slow in late game. Stats depsgraph?

### ISSUE THREAD

**Issue by will-ca** [yairm210/Unciv#5440]
**Platform**
Android

**Version**
3.17.4, Play Store

**Describe the bug**
The build queue and population assignments UIs are very slow for large cities in the late game. It feels like every selected or added item is triggering a full recomputation of the entire empire's population and stats, or something. (I suppose that's probably how it updates the displayed city stats, accounting for all social policies, wonders, etc.)

Technically, the UI itself remains semi-interactive, in that it can sometimes be scrolled and clicked. But the unit/building selector layout element doesn't update until several seconds after the current queue display expands, so there's a broken layout for quite a while. And the scroll position also resets itself once the operation is completely processed— So trying to use it in that time is pointless and frustrating since it will jump around.

This is true for assigning specialists, queuing production, and working tiles.

**Expected behavior**
Expensive calculations should never be done in the middle of interacting with a UI widget. If they must, it should ideally be more asynchronous, so layout and interactivity aren't broken during the computation.

If the expensive calculations are not needed for the interactive display, then they should be deferred and done only once when exiting the city screen IMAO. If they are needed, E.G. to iterate over all buildings in all cities for happiness and thus production penalty... Then I suppose they should be cached at the start of the turn, and recomputed only when something like social policy adoption or building purchase changes them (probably not when reassigning population, and definitely not when doing a UI-only change like adding to the build queue).

**Additional context**
Normally I'd be hesitant to complain about a performance issue, but there does seem to be precedent showing it's accepted here.

Perhaps it would be elegant to implement a dependency graph for game state— Cache all values, and recursively check if their inputs have changed to decide whether to recompute? If it's not already done, it could probably improve performance in more places than just the city screen, while still using lots of dynamic calls that always deliver the most up to date values.

---

I suppose this is actually a single effect with multiple issues behind it:

* [ ] Make UI more asynchronous from empire state, so the City screen stays interactive even with lots of computations.
* [ ] Reduce the amount of computation done during city screen interaction— Cache when entering and defer until exiting where possible.
* [x] Maybe implement game-wide stats cache and dependency graph.

**Comment by K3XRad** [comment]
I can confirm that issue. I even got some crashes with that slow UI when I tried to buy some buildings in a cities. Sorry, but I don't saved the logcat output with those crashes.

**Comment by yairm210** [comment]
Recent performance improvements should have solved this
